waiter/integration/waiter/authentication_test.clj

```diff
@@ -382,3 +384,105 @@
+                                                         http-308-permanent-redirect})
+          assertion-message (str {:headers (:headers authorize-response)
+                                  :status (:status authorize-response)
+                                  :uri authorize-location})
+          {:strs [location]} (:headers authorize-response)]
+      (is (not (str/blank? location)) assertion-message)
+      (if (or (>= iteration 10)
+              (and (str/includes? location "/oidc/v1/callback?")
+                   (str/includes? location "code=")
+                   (str/includes? location "state=")))
+        location
+        (recur (inc iteration) location)))))
+
+(deftest ^:parallel ^:integration-fast test-oidc-authentication-redirect
+  (testing-using-waiter-url
+    (if (oidc-auth-enabled? waiter-url)
+      (let [waiter-host (-> waiter-url sanitize-waiter-url utils/authority->host)
+            waiter-token (or (System/getenv "WAITER_TEST_TOKEN_OIDC")
+                             (create-token-name waiter-url (rand-name)))
+            service-parameters (assoc (kitchen-params)
+                                 :env {"USE_OIDC_AUTH" "true"}
+                                 :name (rand-name)
+                                 :run-as-user (retrieve-username))
+            token-response (post-token waiter-url (assoc service-parameters :token waiter-token))
+            _ (assert-response-status token-response http-200-ok)
+            ;; absence of Negotiate header also triggers an unauthorized response
+            request-headers {"authorization" "SingleUser unauthorized"
+                             "accept-redirect" "yes"
+                             "host" waiter-token
+                             "x-forwarded-proto" "https"}
+            port (waiter-settings-port waiter-url)
+            target-url (str waiter-host ":" port)
+            {:keys [cookies] :as initial-response}
+            (make-request-with-debug-info
+              request-headers
+              #(make-request target-url "/request-info" :disable-auth true :headers % :method :get))]
+        (try
+          (assert-response-status initial-response http-302-moved-temporarily)
+          (let [{:strs [location set-cookie]} (:headers initial-response)
+                assertion-message (str {:headers (:headers initial-response)
+                                        :set-cookie set-cookie
+                                        :status (:status initial-response)})]
+            (is (not (str/blank? location)) assertion-message)
+            (assert-oidc-challenge-cookie set-cookie assertion-message)
+
+            (let [callback-location (follow-authorize-redirects location)]
+              (is (not (str/blank? callback-location)) assertion-message)
+              (is (str/includes? callback-location "/oidc/v1/callback?") assertion-message)
+              (let [callback-uri (URI. (str/replace callback-location #" " "+"))
```

Comment:
Something is off here.  `callback-location` should contain a valid URL that doesn't need any encoding.

Reply:
Addressed this in the JWLS server, the space was being introduced due to the `ring-params/assoc-query-params` invocation there.

https://github.com/twosigma/waiter/pull/1096/files#diff-e9ccf7c23da8998bd4f02073f6a65f23R118-R119